game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java

```diff
@@ -407,69 +418,83 @@ private Territory getRetreatInternal(final String message, final Collection<Terr
-                latch.countDown();
-                return;
-              }
-              // if you have eliminated the impossible, whatever remains, no matter how improbable,
-              // must be the truth
-              // retreat
-              if (possible.size() == 1) {
-                retreatTo.set(possible.iterator().next());
+              actionButton.setEnabled(false);
+              if (showRetreatDialog(message, possible, retreatTo)) {
+                actionButton.setAction(nullAction);
                 latch.countDown();
-              } else {
-                final RetreatComponent comp = new RetreatComponent(possible);
-                final int option =
-                    JOptionPane.showConfirmDialog(
-                        BattleDisplay.this,
-                        comp,
-                        message,
-                        JOptionPane.OK_CANCEL_OPTION,
-                        JOptionPane.PLAIN_MESSAGE,
-                        null);
-                if (option == JOptionPane.OK_OPTION) {
-                  if (comp.getSelection() != null) {
-                    retreatTo.set(comp.getSelection());
-                    latch.countDown();
-                  }
-                }
               }
+              actionButton.setEnabled(true);
             });
-    SwingUtilities.invokeLater(() -> actionButton.setAction(action));
-    SwingUtilities.invokeLater(() -> action.actionPerformed(null));
+    SwingUtilities.invokeLater(
+        () -> {
+          actionButton.setAction(action);
+          action.actionPerformed(null);
+        });
     mapPanel.getUiContext().addShutdownLatch(latch);
     Interruptibles.await(latch);
     mapPanel.getUiContext().removeShutdownLatch(latch);
-    SwingUtilities.invokeLater(() -> actionButton.setAction(nullAction));
+
     return retreatTo.get();
   }
 
+  private boolean showRetreatDialog(
+      final String message,
+      final Collection<Territory> possible,
+      final AtomicReference<Territory> retreatTo) {
+    final String yes =
+        possible.size() == 1 ? "Retreat to " + possible.iterator().next().getName() : "Retreat";
+    final String no = "Remain";
+    final String cancel = "Ask Me Later";
+    final String[] options = {yes, no, cancel};
+    final int choice =
+        JOptionPane.showOptionDialog(
+            BattleDisplay.this,
+            message,
+            "Retreat?",
+            JOptionPane.YES_NO_CANCEL_OPTION,
+            JOptionPane.PLAIN_MESSAGE,
+            null,
+            options,
+            no);
+    // dialog dismissed
+    if (choice == JOptionPane.CLOSED_OPTION) {
+      return false;
+    }
+    // wait
+    if (choice == JOptionPane.CANCEL_OPTION) {
+      return false;
+    }
+    // remain
+    if (choice == JOptionPane.NO_OPTION) {
+      return true;
+    }
+    // if you have eliminated the impossible, whatever remains, no matter how improbable,
+    // must be the truth
```

Comment:
This is a great sherlock holmes quote, but this should be remove because I don't get what it's trying to imply? 😅

Reply:
Done.